### PR TITLE
Draft: supporting API minor versions (Do not submit)

### PR DESCRIPTION
**Pull request by adisuissa**
Signed-off-by: Adi Suissa-Peleg <adip@google.com>

This is a draft PR based on the proposal in #8416, allowing to gather comments on whether we are heading in the right direction.
The plan is to gather some feedback, and introduce incremental PRs.

Roadmap:
1. Introduce API version (X.Y.Z) & accessors  - current draft (API_VERSION, source/common/version/api_version.h)
2. Introduce proto’s deprecated fields/enum minor_version annotation - current draft + example (api/envoy/annotations/deprecation.proto)
3. API feature life-cycle description - current draft (api/API_VERSIONING.md)
4. Adding minor version range as context params - current draft (source/common/config/context_provider_impl.h)
5. Adding DiscoveryResponseStatus - current draft (api/envoy/service/discovery/v3/discovery.proto)
6. Test showing the minor version range request - current draft (ads_integration_test.cc)
7. Adding patch version to a (Delta)DiscoveryRequest - current draft(api/envoy/service/discovery/v3/discovery.proto)
8. Example of a test using minor version - current draft (ads_integration_test.cc)
9. Example of a test using patch version - future work
10. Introduce stats counters for failed resource queries due to versioning - future work
11. Manifest file design doc - future work
12. Manifest file implementation - future work
13. Tools supporting minor version updates - future work
14. Tools supporting patch version updates - future work
15. Tools for Manifest file updates - future work

### CODE REVIEW

**Comment by adisuissa** [comment]
@envoyproxy/api-shepherds  PTAL

**Comment by htuch** [comment]
I'll take a first pass.

File: api/API_VERSIONING.md [hunk 1]
```diff
@@ -28,9 +29,24 @@ configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). The
 v3 API bootstrap configuration is `envoy.config.bootstrap.v3.Boostrap`, even
 though it might might transitively reference `envoy.service.trace.v2`.
 
+The minor version enables API stability and support, and to safely deprecate old
+features and reduce the tech debt. A client can support a range of minor API
+versions, and can request xDS resources that match this range, using the resource
+[context parameters](https://github.com/cncf/udpa/blob/master/xds/core/v3/context_params.proto).
```
Comment (htuch) [inline api/API_VERSIONING.md:35]: Also `Node` will indicate for simple string names. I'd generally lead in with the old way of doing it and then point out the new way.

File: api/API_VERSIONING.md [hunk 2]
```diff
@@ -28,9 +29,24 @@ configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). The
...
+Deprecated features, from previous minor or major versions (and their code), can be safely removed
+from the client. In general, within a minor API version, we do not allow any breaking changes
+(more details below). The minor version will be reset to 0 upon a release of a new major version.
```
Comment (htuch) [inline api/API_VERSIONING.md:38]: "any breaking changes to the Protobuf definitions for the API".

File: api/API_VERSIONING.md [hunk 3]
```diff
@@ -28,9 +29,24 @@ configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). The
...
+
+The patch version tracks incremental changes to the API, such as, adding a new
+package, adding a new field, deprecating a field, etc.
+A client can supply the supported patch version when requesting xDS resources,
+which allows clients and servers to support new features, before the release of
```
Comment (htuch) [inline api/API_VERSIONING.md:43]: I'd be more crisp. It's already possible for servers to unconditionally set new fields and have some clients support and others ignore. What's not possible today is for a server to detect if a client supports a newer field without patch version. This is a way for servers to selectively condition there response based on whether a client has a new feature or not.

File: api/API_VERSIONING.md [hunk 4]
```diff
@@ -28,9 +29,24 @@ configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). The
...
+the next minor version.
+The patch version will be reset to 0 upon a release of a new minor version.
+
 # Backwards compatibility
 
-In general, within a package's major API version, we do not allow any breaking changes. The guiding
+In general, within a minor API version, we do not allow any breaking changes. The guiding
```
Comment (htuch) [inline api/API_VERSIONING.md:49]: "major and minor".

File: api/API_VERSIONING.md [hunk 5]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
 management servers to xDS-as-a-service offerings run by vendors. The [xDS API
 shepherds](https://github.com/orgs/envoyproxy/teams/api-shepherds) will make the decision to add a
 new major version subject to the following constraints:
-* There exists sufficient technical debt in the xDS APIs in the existing supported major version
+* There exists a major, breaking change, to the xDS APIs in the existing supported major version
```
Comment (htuch) [inline api/API_VERSIONING.md:103]: I don't think this needs to change in this PR. The criteria for v4 (if ever) should be discussed, but it's unrelated to adding minor versions.

File: api/API_VERSIONING.md [hunk 6]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
...
   to justify the cost burden for xDS client/server implementations.
 * At least one year has elapsed since the last major version was cut.
 * Consultation with the Envoy community (via Envoy community call, `#xds` channel on Slack), as
   well as gRPC OSS community (via reaching out to language maintainers) is made. This is not a veto
   process; the API shepherds retain the right to move forward with a new major API version after
   weighing this input with the first two considerations above.
 
-Following the release of a new major version, the API lifecycle follows a deprecation clock.
```
Comment (htuch) [inline api/API_VERSIONING.md:95]: I don't think we should delete this yet, see above comment. Instead, create a subsection header for major version clock and another for minor version (which is what you're adding below).

File: api/API_VERSIONING.md [hunk 7]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
...
-retain implementation support for at least 1-2 years.
-
-We are currently working on a strategy to introduce minor versions
-(https://github.com/envoyproxy/envoy/issues/8416). This will bump the xDS API minor version on every
-deprecation and field introduction/modification. This will provide an opportunity for the control
-plane to condition on client and major/minor API version support. Currently under discussion, but
-not finalized will be the sunsetting of Envoy client support for deprecated features after a year
-of support within a major version. Please post to https://github.com/envoyproxy/envoy/issues/8416
-any thoughts around this.
+A new API minor version is expected to be released once a year, around Q1. This
```
Comment (htuch) [inline api/API_VERSIONING.md:111]: This is going to be "at the Q2 release" as things stand.

File: api/API_VERSIONING.md [hunk 8]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
...
+allows for gradual introduction of new capabilities and features that will be supported
+by clients, and by control planes.
+Following the release of a new minor version, the API lifecycle follows a deprecation clock.
+Envoy will support the most recent 3 minor versions, including the current. This
```
Comment (htuch) [inline api/API_VERSIONING.md:115]: I think it's the most recent 2 minor versions, including the current. I.e. In 2022 Q2 we're on 3.2.0, we still support 3.1.0 for another year and we remove 3.0.0 deprecated fields.

File: api/API_VERSIONING.md [hunk 9]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
...
+as adding a new enum value, or setting a field as deprecated. We expect frequent
+updates to the patch version.
+
+The following is a an example of the API versioning lifecycle. Assume that the
+current API version is v3.7.18. Assume that a new change (pull-request) updates the API as follows:
+`field1` is annotated as deprecated and `field2`, a new field, is added.
+As a consequence of this, the patch version is increased and the API version becomes v3.7.19.
+Upon sending a request for xDS resources, the client provides the resource type
+which includes its full package path which entails the API major version (v3), and
+the range of minor versions that the client supports (5 to 7) using the resource
```
Comment (htuch) [inline api/API_VERSIONING.md:129]: 6 to 7

File: api/API_VERSIONING.md [hunk 10]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
...
+context parameters: `xds.api.oldest_minor_ver=5` and `xds.api.latest_minor_ver=7`.
```
Comment (htuch) [inline api/API_VERSIONING.md:130]: As above, best to have both Node and context parameter examples as many will use Node.

File: api/API_VERSIONING.md [hunk 11]
```diff
@@ -84,52 +100,43 @@ from clients (Envoy, gRPC) and a large number of control planes, ranging from si
...
+In addition, the client will provide patch version (19) as part of the xDS resource request.
+This allows the control plane to send either resources with `field1` or with `field2`.
+In the next year's Q1, the current API minor version will be cut, and the API
+version becomes v3.8.0. The client will now support a new minor versions range
+(6 to 8), and features deprecated in minor version 5 will be removed.
```
Comment (htuch) [inline api/API_VERSIONING.md:135]: 7 to 8

File: api/API_VERSIONING.md [hunk 12]
```diff
@@ -144,6 +151,18 @@ only be made to the *current stable major version*. The rationale for this polic
...
+As previously discussed, the new feature is added to the current minor version, and increases the
+patch version.
+
+# API feature deprecation
+
+A feature can be deprecated by annotating the corresponding field or enum value as `[deprecated = true]`.
+This implies that the feature is deprecated during the current minor version, `Y`.
+When the minor version is cut (starting with minor version `Y+1`), automatic tools will add another annotation,
+`[deprecated = true, minor_ver=Y]` (or `minor_ver_enum` for an enum value), indicating that the field
+(or enum value) were deprecated during minor version Y. 2 years later, when minor version `Y+2` will
```
Comment (htuch) [inline api/API_VERSIONING.md:163]: 1-2 years later when minor version Y+1..

File: api/envoy/annotations/deprecation.proto [hunk 13]
```diff
@@ -5,18 +5,24 @@ package envoy.annotations;
...
+// disallowed by default, a state which is reversible with
+// :ref:`runtime overrides <config_runtime_deprecation>`.
 
 // Magic number in this file derived from top 28bit of SHA256 digest of
-// "envoy.annotation.disallowed_by_default"
+// "envoy.annotation.disallowed_by_default" and "envoy.annotation.minor_ver"
 extend google.protobuf.FieldOptions {
   bool disallowed_by_default = 189503207;
+
+  uint32 minor_ver = 240115895;
```
Comment (htuch) [inline api/envoy/annotations/deprecation.proto:19]: Given that things are already pretty verbose, e.g. `envoy.annotation.minor_ver`, it might be clearer to make it the full `envoy.annotation.deprecated_at_minor_ver`.

File: api/envoy/config/core/v3/deprecaetd_example.proto [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+option java_outer_classname = "DeprecaetdExampleProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Test message]
+
+// EnumExample
+enum EnumExample {
+  // Enum value deprecated in minor version 5.
+  VAL1 = 0 [deprecated = true, (envoy.annotations.minor_ver_enum) = 5];
```
Comment (htuch) [inline api/envoy/config/core/v3/deprecaetd_example.proto:19]: This is technically fully expressive, but might be clearer if it looked like "3.5" to the reader. Not sure if this is actionable.

File: api/envoy/service/discovery/v3/discovery.proto [hunk 15]
```diff
@@ -107,6 +128,22 @@ message DiscoveryResponse {
...
   // The control plane instance that sent the response.
   config.core.v3.ControlPlane control_plane = 6;
+
+  // [#not-implemented-hide:]
+  // An optional per-resource status response that can be used when no matching resource(s)
+  // are sent to the client, and will describe the reason. This will be a map
+  // from a resource name to its status. A resource that doesn't appear in the map
+  // implies the resource response is valid. No status map implies that all the
+  // resources in the response are valid.
+  map<string, DiscoveryResponseStatus> resources_status = 7;
```
Comment (htuch) [inline api/envoy/service/discovery/v3/discovery.proto:138]: As discussed IRL, we should probably move this to a separate PR for this feature in general. @markdroth would this help get rid of the "15 second rule"?

File: api/envoy/service/discovery/v3/discovery.proto [hunk 16]
```diff
@@ -107,6 +128,22 @@ message DiscoveryResponse {
...
+
+  // [#not-implemented-hide:]
+  // An optional API patch version of the requested resource that is supported by
+  // the client. This is useful for use-cases where the client and server support new
+  // API features, before the next API minor version is rolled out.
+  // If not included, the patch version defaults to 0, stating that the client
+  // supports API features up to the client's minor version.
+  uint32 patch_version = 8;
```
Comment (htuch) [inline api/envoy/service/discovery/v3/discovery.proto:146]: Why is this in response? Shouldn't it only be request?

**Review by htuch — CHANGES_REQUESTED** [review]
Awesome, this is great to see coming to fruition. I've left some high level design comments. @markdroth probably should also take a pass (in particular around transport-level changes) as a next round.
/wait

File: API_VERSION [hunk 17]
```diff
@@ -0,0 +1 @@
+3.7.0
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply API_VERSION:1]: I agree that the first minor version will be 3.1.0, and all deprecated fields will have the deprecated_at annotation set to 3.0.
At the moment it is set to 3.7.0 to "comply" with the api/envoy/config/core/v3/deprecaetd_example.proto example contents.

File: api/API_VERSIONING.md (hunk 1 above)
Comment (adisuissa) [reply api/API_VERSIONING.md:35]: Added `user_agent_latest_api_version` and `user_agent_oldest_api_version` to `Node`.

File: api/API_VERSIONING.md (hunk 2 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Done.

File: api/API_VERSIONING.md (hunk 3 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Rephrased.

File: api/API_VERSIONING.md (hunk 4 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Done.

File: api/API_VERSIONING.md (hunk 5 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Reverted.

File: api/API_VERSIONING.md (hunk 7 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Done.

File: api/API_VERSIONING.md (hunk 8 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Yes, I misunderstood the original requirement.
This means that there may be features that their support will need to be updated in 1 year time (+epsilon).

File: api/API_VERSIONING.md (hunk 9 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Fixed.

File: api/API_VERSIONING.md (hunk 10 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Done.

File: api/API_VERSIONING.md (hunk 11 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Done.

File: api/API_VERSIONING.md (hunk 12 above)
Comment (adisuissa) [reply api/API_VERSIONING.md]: Rephrased and updated.

File: api/envoy/annotations/deprecation.proto (hunk 13 above)
Comment (adisuissa) [reply api/envoy/annotations/deprecation.proto]: Done.

File: api/envoy/config/core/v3/deprecaetd_example.proto (hunk 14 above)
Comment (adisuissa) [reply api/envoy/config/core/v3/deprecaetd_example.proto]: Done.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 16 above)
Comment (adisuissa) [reply api/envoy/service/discovery/v3/discovery.proto]: Moved to `DiscoveryResponse`.

**Comment by adisuissa** [comment]
A description of the tools needed for minor/patch versioning verifications and updates (without the Manifest tools):
https://docs.google.com/document/d/1cZjuqBp8lIY4vOzppauFpexvRahtkRKpFlpEG0uQRkI/edit?usp=sharing

File: source/common/version/api_version.cc [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+const envoy::config::core::v3::ApiVersionNumber& ApiVersionInfo::oldestApiVersion() {
+  static const auto* result =
+      new envoy::config::core::v3::ApiVersionNumber(computeOldestApiVersion(apiVersion()));
+  return *result;
+}
+
+envoy::config::core::v3::ApiVersionNumber ApiVersionInfo::makeApiVersion(const char* version) {
+  envoy::config::core::v3::ApiVersionNumber result;
+  // Split API_VERSION_NUMBER into version
+  std::regex ver_regex("([\\d]+)\\.([\\d]+)\\.([\\d]+)");
```
Comment (htuch) [inline source/common/version/api_version.cc:35]: Nit: maybe just `absl::StrSplit` on `.`

**Review by htuch — DISMISSED** [review]
Looks great, I've left another round of feedback. LMK when to take another pass. Thanks.
/wait

File: api/envoy/service/discovery/v3/discovery.proto (hunk 15 above)
Comment (markdroth) [reply api/envoy/service/discovery/v3/discovery.proto:145]: Yeah, I think something like this is a good idea.  But I do think it should probably be a separate PR.

Do we actually want this to be a separate map?  An alternative would be to embed the `DiscoveryResponseStatus` inside of the `Resource` message in the normal `resources` field.  That way, clients just need to look at one field for every resource, and some of those resources might be a special type that indicates that the server didn't have the requested resource.

Also, what is the expected semantic here when a subscribed resource does not exist and multiple responses are sent during the time it does not exist?  For LDS and CDS, is the server obligated to send every non-existent resource in every response, just like it normally does?  For other resource types, can the server send the response status just once, and the client assumes that the status doesn't change until/unless it receives a subsequent response that either populates the resource or changes its response status?  (The semantic here might be a bit more obvious if the `DiscoveryResponseStatus` was inside of the `resources` field instead of being a separate map.)

File: api/API_VERSIONING.md [hunk 19]
```diff
@@ -28,11 +29,31 @@ configuration resources (e.g. bootstrap, xDS resources such as `Cluster`). The
 v3 API bootstrap configuration is `envoy.config.bootstrap.v3.Boostrap`, even
 though it might might transitively reference `envoy.service.trace.v2`.
 
+The minor version enables API stability and support, and to safely deprecate old
+features and reduce the tech debt. A client can support a range of minor API
+versions, and can request xDS resources that match this range, using the resource
+[context parameters](https://github.com/cncf/udpa/blob/master/xds/core/v3/context_params.proto).
```
Comment (markdroth) [inline api/API_VERSIONING.md:35]: I would still like to consider adding a flexible range-based matching semantic for minor versions.  Can we do that as part of this effort, or do we want to decouple that from this?

Hmm...  Actually, now that we've decided to stick with string representations for new-style resource names, doesn't that mean that we don't actually have a way to add more flexible matching syntax later if we don't do it now?  When we were going to be using `ResourceLocator` on the wire, it would have been easy to add new matching types, because they'd be expressed in different fields.  But with the string form, it's basically just "key=value&key=value&...".  So if we're interpreting the values literally, then any new special syntax we add there may break existing users.

It seems like we need to either add the more flexible matching syntax now or carve out some sort of escape hatch in the URI syntax to allow for future changes.  (Actually, the latter would probably be a good idea even if we do the former, since we may want to add other flexible matcher types in the future.)

File: api/envoy/service/discovery/v3/discovery.proto [hunk 20]
```diff
@@ -62,9 +62,38 @@ message DiscoveryRequest {
...
   // debugging, the string provided is not guaranteed to be stable across Envoy versions.
   google.rpc.Status error_detail = 6;
+
+  // [#not-implemented-hide:]
+  // Latest xDS API patch version supported by the client.
+  // This is useful for use-cases where the client and server support new
+  // API features, before the next API minor version is rolled out.
+  // If not included, the patch version defaults to 0, stating that the client
+  // supports API features up to the client's minor version.
+  uint32 patch_version = 7;
```
Comment (markdroth) [inline api/envoy/service/discovery/v3/discovery.proto:72]: Why is this here instead of being in the `Node` message?  As currently designed, it's by definition something that works only when directly talking to an authoritative server, because it's not part of the resource name, so caching proxies can't handle it.

Note that if we do add a mechanism for flexible range-based context param matching, then we can make this part of the resource name, so it would work basically the same way as the minor version.  I think that would actually make this mechanism easier for people to understand.

File: api/envoy/service/discovery/v3/discovery.proto [hunk 21]
```diff
@@ -230,6 +275,14 @@ message DeltaDiscoveryResponse {
...
   // The control plane instance that sent the response.
   config.core.v3.ControlPlane control_plane = 7;
+
+  // [#not-implemented-hide:]
+  // An optional per-resource status response that can be used when no matching resource(s)
+  // are sent to the client, and will describe the reason. This will be a map
+  // from a resource name to its status. A resource that doesn't appear in the map
+  // implies the resource response is valid. No status map implies that all the
+  // resources in the response are valid.
+  map<string, DiscoveryResponseStatus> resources_status = 8;
```
Comment (markdroth) [inline api/envoy/service/discovery/v3/discovery.proto:285]: Note that unlike the SotW protocol, which currently doesn't have a way to explicitly indicate that a subscribed resource does not exist, the delta protocol *does* have such a mechanism (the `removed_resources` field).  So in this case, the only value of this field is to support the version-mismatch case.  And if we handle the version via flexible range-based context parameter matching, then we wouldn't need this field at all.

**Review by markdroth — COMMENTED** [review]
Sorry for taking so long to review this; I've been pretty busy lately.

It might be a good idea for us to sync up in person to decide the best way to proceed with this.  I'd still prefer to consider handling this via flexible range-based context parameter matching if possible.

Please let me know if you have any questions.  Thanks!

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:285]: > And if we handle the version via flexible range-based context parameter matching, then we wouldn't need this field at all.

How does this relate? The management server can still be missing a version in the specified range.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:72]: If it's not federated, it can live in the `Node` message agreed.

File: api/API_VERSIONING.md (hunk 19 above)
Comment (htuch) [reply api/API_VERSIONING.md:35]: I'm on the fence on this one. I would like to future proof with some notion of escape hatch in the URI syntax, that's a no brainer. I think I'd like to understand the full set of flexible match objectives before deciding on a concrete syntax and what do use here instead of min/max version. How do you think is best to go forward on that @markdroth?

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (markdroth) [reply api/envoy/service/discovery/v3/discovery.proto:285]: Right, but if the version was just a context parameter, then it would effectively be part of the name, so `NO_MATCHING_RESOURCE` and `NO_MATCHING_VERSION_RESOURCE` would be the same thing.  And at that point, we would not need this field, because the existing `removed_resources` field would do what we need.

Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:285]: In this case, we've actually strictly lost information, but I agree it's a more consistent and logical way of viewing things. I think @mattklein123 was interested in a specific "no matching version" feedback signal, but maybe this would be sufficient.

File: api/API_VERSIONING.md (hunk 19 above)
Comment (adisuissa) [reply api/API_VERSIONING.md:35]: I agree that range-based matching semantics parameters should be supported, but I'm not certain I understand many of the use-cases that it will be used for.
I think that eventually we can replace this with a parameter similar to this: `xds.api.range.minor_version={"start":5,"end":6}`.

Comment (mattklein123) [reply api/API_VERSIONING.md:35]: Is the concern here that we don't want magic params that have prior agree to semantics? Meaning, if we have an oldest value and a newest value it seems not hard to just do the range matching manually.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:72]: I'm probably missing something but why isn't it effectively part of the resource name? Meaning, if we assume the patch level for a specific message is a recursive walk of all sub-messages, don't we know it on a per message/resource basis and then it might be different and should be set here? I guess maybe I'm then also confused why we don't need major/minor here also? Or we assume that those are sent in context params?

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:285]: I think the requirement here is that we have a clear way of providing status to the client about why a resource was not returned. The `DiscoveryResponseStatus` message seems to do that nicely. I don't have a strong preference on how we do it exactly but I agree with @htuch that we should not lose information on why something happened and should probably have it look similar for both SoTW and delta, whatever the best way of accomplishing that is.

**Review by mattklein123 — COMMENTED** [review]
Very exciting to see this happening. A few high level comments/questions.

File: api/API_VERSIONING.md (hunk 19 above)
Comment (adisuissa) [reply api/API_VERSIONING.md:35]: IIUC the issue of exact-matching is that the returned resource must have the same parameters.
Range-parameters enable explicit semantics. For example, a cache (or xds-relay) will be able to return the same resource with minor version 5 for a request with the range [4,5] and for a different request with the range [5,6].

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (adisuissa) [reply api/envoy/service/discovery/v3/discovery.proto:72]: I think it's related to the semantics of exact-match and of an optional patch support.
If this is part of the context-params, caches should only return resources with the patch, and caches will need to cache the same resource with different patch versions.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (markdroth) [reply api/envoy/service/discovery/v3/discovery.proto:285]: To be clear, I certainly don't object to having a way of exposing a more detailed status to the client; I think that's definitely a good thing for debuggability.  I'm more concerned about having two different ways of doing the same thing.  An alternative here would be to simply deprecate the existing `removed_resources` field and replace it with this.

Also, could we consider using `google.rpc.Status` instead of a custom `DiscoveryResponseStatus` enum?  That seems a bit more flexible in the long run.

File: api/API_VERSIONING.md (hunk 19 above)
Comment (markdroth) [reply api/API_VERSIONING.md:35]: Right, the issue is that the current exact-match semantic means that the context params in the result must exactly match the ones in the request.  So if you have one client requesting `xds.api.range.minor_version=[0, 1]` and another requesting `xds.api.range.minor_version=[1, 2]`, you can provide the contents at version 1 to both of them, but you need to return each one with a different context parameter.  This will basically lead to cache pollution, where any caching proxy will need an increasing number of copies of each identical resource under different names, and the pollution will increase combinatorically.

What I am proposing is a range-based matching mechanism where the clients can ask for something like `xds.api.range.minor_version=SPECIAL_RANGE_SYNTAX[0, 1]` and `xds.api.range.minor_version=SPECIAL_RANGE_SYNTAX[1, 2]`, and both can get back the same resource with context parameter `xds.api.range.minor_version=1`.  In order to make that work, we need to define the syntax for range-based matching, make sure it doesn't conflict with any "flat" context parameter values (e.g., by defining some sort of escaping mechanism), and make sure that xDS servers, clients, and proxies understand it.

This is a bit more work, but I think it's also a cleaner solution.  And I think this same mechanism will be able to be used for other use-cases in the future.  (We already have one use-case internally for which we've been discussing something like this.)

@htuch, assuming that we have a reasonable way to future-proof the syntax in a way that would allow us to add more flexible matching types if we need to do so in the future, I think I'd advocate for restricting our initial changes to just the cases that we need immediately.  For minor version, I think that's range-based matching.  And if we want to handle the patch version the same way, we might also want a simple wildcard match for that.  I think anything else can wait for the future.

I would be happy to write up a proposal for how this would work, but I need something to build off of.  Where did we leave things on plans for the initial xRFC for the new naming scheme?  Are you still planning to do that at some point?

Comment (mattklein123) [reply api/API_VERSIONING.md:35]: OK thanks this makes sense to me. I agree we should do the work to implement range syntax. It doesn't seem too terribly difficult.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:72]: OK thanks, this makes sense. Both this convo and the range convo does make me wonder a bit if we are thinking about context params wrong and the server should actually be able to return a result with a subset of matching params? Does that make sense?

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:285]: I'm fine with whatever solution you all agree to on this as long as we maintain the data fidelity.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (markdroth) [reply api/envoy/service/discovery/v3/discovery.proto:72]: I think the core problem is that clients need a way to know which resources correspond to which of their requests.  If the server can send back resources that are just missing some of the params that the client ask for, then there's no way to unambiguously map the resources to the requests.

For example, let's say that a client has subscribed to the following:
- `xdstp:///envoy.config.route.v3.RouteConfiguration/route_config?my_context_param=foo`
- `xdstp:///envoy.config.route.v3.RouteConfiguration/route_config?my_context_param=bar`

If the server sends back a resource whose name is `xdstp:///envoy.config.route.v3.RouteConfiguration/route_config` with no `my_context_param` parameter, then how can the client know which of its requests that is the response to?

This is why I think that we're going to need clearly defined flexible matching semantics.  We need to be very sure that we don't accidentally introduce any ambiguity here.

File: api/API_VERSIONING.md (hunk 19 above)
Comment (htuch) [reply api/API_VERSIONING.md:35]: Yes, a proposal here for a syntax which is future proof and can initially accommodate simple range would help unblock work here I think. Thanks Mark.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:72]: Yes, let's do range and move things to patch, thanks.

Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:72]: What if there was a special wildcard param value? (Just thinking out loud here). This would then allow intermediate caches to be quite intelligent about mappings. Taking the patch value this would allow the server to respond with `patch_value=*` or whatever.

Comment (adisuissa) [reply api/envoy/service/discovery/v3/discovery.proto:72]: I agree with moving patch into `Node`. Actually `Node` already has a field `ApiVersionNumber user_agent_latest_api_version` where a client can add the patch version it supports.
Just wanted to note that at the moment there's no support for patch version in the xdstp naming scheme version negotiation, and that should be fine because patch version support is optional.

> What if there was a special wildcard param value? (Just thinking out loud here). This would then allow intermediate caches to be quite intelligent about mappings. Taking the patch value this would allow the server to respond with `patch_value=*` or whatever.

IMO the _patch_ field should actually be understood as _maximal supported patch version_, so if the server responds with `patch_value=*`, the intermediate servers will not be able to know if the resource is supported in `max_patch_version=3` or in `max_patch_version=6`, and they'll need that to process future requests.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 15 above)
Comment (adisuissa) [reply api/envoy/service/discovery/v3/discovery.proto:145]: I agree that it makes more sense to add the `DiscoveryResponseStatus` to each specific resource (in `Resource`).
I'm probably missing something here, but IIUC the `resources` field in a SotW response is of type `google.protobuf.Any`, and the response is not required to return the resources in a wrapped `Resource`. If this implies that the `Resource` wrapper is required how do we plan to the gradual introduction of this to existing xDS services?

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (adisuissa) [reply api/envoy/service/discovery/v3/discovery.proto:285]: > Right, but if the version was just a context parameter, then it would effectively be part of the name, so `NO_MATCHING_RESOURCE` and `NO_MATCHING_VERSION_RESOURCE` would be the same thing. And at that point, we would not need this field, because the existing `removed_resources` field would do what we need.

Maybe I got it wrong, but I thought of the response status as a complimentary field to the removed resources, providing the reason why were they removed. It essentially allows the following differentiation: a requested resource name (no matter what version) doesn't exist in the management server; and a requested resource in a specific version range doesn't exist in the management server (the server only handles older vs newer versions).
At the moment, the client doesn't use this information (other than logging) but it does allow the flexibility to request the resource from other servers, or to notify the operator that there's an issue that needs to be addressed (the client needs to be updated).

 > Also, could we consider using `google.rpc.Status` instead of a custom `DiscoveryResponseStatus` enum? That seems a bit more flexible in the long run.

I agree it should be used, and stating that the _code_ field in it will be one of the `DiscoveryResponseStatusTypes` that we define.

Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:285]: I don't feel strongly about it, but I tend to agree with Mark that resources_status could be a superset of removed_resources and we can just build it in a way that would allow differentiating the two cases?

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (mattklein123) [reply api/envoy/service/discovery/v3/discovery.proto:72]: > , the intermediate servers will not be able to know if the resource is supported in max_patch_version=3 or in max_patch_version=6, and they'll need that to process future requests.

But isn't that effectively by design? The management server shouldn't send * unless it really means it. I think what I'm getting at is it would give a way for the management server to effectively ignore patch that is cache compatible.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 21 above)
Comment (markdroth) [reply api/envoy/service/discovery/v3/discovery.proto:285]: > At the moment, the client doesn't use this information (other than logging) but it does allow the flexibility to request the resource from other servers, or to notify the operator that there's an issue that needs to be addressed (the client needs to be updated).

I suspect that there are no cases where the client will need to programmatically differentiate between these two cases.  Knowing the difference is useful for humans but not really helpful for client implementations.  Most clients will simply treat the resource as non-existing in both cases; I could imagine a client deciding to try another management server instead, but it seems unlikely that a client would choose to do so in one of these two cases but not the other.

That's why I favor simply deprecating the existing `removed_resources` field and replacing it with `map<string, google.rpc.Status> resources_not_available`, where the map value would be an error message indicating why the specified resource was not available.  This approach would allow us to handle any arbitrary reason why the resource is not available, in case we run into cases other than the two we are currently considering.

> I agree it should be used, and stating that the _code_ field in it will be one of the `DiscoveryResponseStatusTypes` that we define.

I think the code in that proto needs to be one of the standard status code enum values.  It would be very surprising to users for the code to be a new error space.

Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:285]: Yeah, the primary requirement is debuggability; as long as we can include the version failure in the `google.rpc.Status` string it's all good.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 20 above)
Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:72]: I think it's possible to introduce a notion of cache-ignored context params. This is analogous to the ability in CDNs to ignore a list of query params. We've discussed adding this a few times to the spec.. I think the point becomes moot if we have range params though, since instead of max supported version N we can just say `[0..N]` with the same semantics and caches basically ignore it.
